modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/type/IgniteTypeFactorySelfTest.java

```diff
@@ -0,0 +1,170 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.sql.engine.type;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Random;
+import java.util.stream.Stream;
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.sql.type.SqlTypeName;
+import org.apache.ignite.internal.sql.engine.util.Commons;
+import org.apache.ignite.internal.testframework.BaseIgniteAbstractTest;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+
+/**
+ * Tests for {@link IgniteTypeFactory}.
+ */
+public class IgniteTypeFactorySelfTest extends BaseIgniteAbstractTest {
+
+    private static final IgniteTypeFactory TYPE_FACTORY = Commons.typeFactory();
+
+    private final Random random = new Random();
+
+    @BeforeEach
+    public void test() {
+        long seed = System.nanoTime();
+        random.setSeed(seed);
+        log.info("Seed: {}", seed);
+    }
+
+    @ParameterizedTest
+    @MethodSource("types")
+    public void leastRestrictiveTypeBetweenTimestampsTypes(List<RelDataType> types, RelDataType expected) {
+        RelDataType actual = TYPE_FACTORY.leastRestrictive(types);
+
+        List<RelDataType> shuffledTypes = new ArrayList<>(types);
+        Collections.shuffle(shuffledTypes, random);
+
+        assertEquals(expected, actual);
+    }
+
+    private static Stream<Arguments> types() {
```

Comment:
let's add a few test cases on permutation of `timestamp`, `timestampLtz`, and `NULL`-literal (`SqlTypeName.NULL`)